csharp/src/Ice/IceDiscovery/LocatorRegistry.cs

```diff
@@ -0,0 +1,298 @@
+                    var endpoints = adapterIds.SelectMany(id => _ice1Adapters[id].Endpoints).ToList();
+                    return (_dummyIce1Proxy.Clone(endpoints: endpoints), true);
+                }
+
+                return (null, false);
+            }
+        }
+
+        internal IObjectPrx? FindObject(Identity identity)
+        {
+            lock (_mutex)
+            {
+                if (identity.Name.Length == 0)
+                {
+                    return null;
+                }
+
+                foreach ((string replicaGroupId, Protocol protocol) in _replicaGroups.Keys)
+                {
+                    if (protocol == Protocol.Ice1)
+                    {
+                        try
+                        {
+                            // This proxy is an indirect proxy with a location (the replica group ID).
+                            IObjectPrx proxy = _dummyIce1Proxy.Clone(
+                                IObjectPrx.Factory,
+                                identity: identity,
+                                location: ImmutableArray.Create(replicaGroupId));
+                            proxy.IcePing();
+                            return proxy;
+                        }
+                        catch
+                        {
+                            // Ignore and move on to the next replica group
+                        }
+                    }
+                }
+
+                foreach (string adapterId in _ice1Adapters.Keys)
+                {
+                    try
+                    {
+                        IObjectPrx proxy = _dummyIce1Proxy.Clone(
+                            IObjectPrx.Factory,
+                            identity: identity,
+                            location: ImmutableArray.Create(adapterId));
+                        proxy.IcePing();
+                        return proxy;
+                    }
+                    catch
+                    {
+                        // Ignore.
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        internal (IReadOnlyList<EndpointData> Endpoints, bool IsReplicaGroup) ResolveAdapterId(string adapterId)
+        {
+            lock (_mutex)
+            {
+                if (_ice2Adapters.TryGetValue(adapterId, out IReadOnlyList<EndpointData>? endpoints))
+                {
+                    return (endpoints, false);
+                }
+
+                if (_replicaGroups.TryGetValue((adapterId, Protocol.Ice2), out HashSet<string>? adapterIds))
+                {
+                    Debug.Assert(adapterIds.Count > 0);
+                    return (adapterIds.SelectMany(id => _ice2Adapters[id]).ToList(), true);
+                }
+
+                return (ImmutableArray<EndpointData>.Empty, false);
+            }
+        }
+
+        internal (IReadOnlyList<EndpointData> Endpoints, string AdapterId) ResolveWellKnownProxy(Identity identity)
```

Comment:
It would be good to change this method to be async and replace the synchronous IcePing call with an async IcePing.

Reply:
It sounds good but it makes the implementation more difficult as we cannot lock _mutex while making this IcePingAsync call. Since we are attempting to ping a local object, is it worth the trouble to copy the keys / adapter IDs?

Reply:
Fixed as suggested.